onnx/onnx.in.proto

```diff
@@ -442,6 +447,21 @@ message TensorProto {
   repeated uint64 uint64_data = 11 [packed = true];
 }
 
+// A serialized sparse-tensor value
+message SparseTensorProto {
+  // The sequence of non-default values are encoded as a tensor of shape [NNZ].
+  // The default-value is zero for numeric tensors, and empty-string for string tensors.
+  optional TensorProto values = 1;
+
+  // The indices of the non-default values are encoded as a tensor of shape [NNZ].
+  // Note that the indices are linearized-indices. Each linearized index can be
+  // converted into a tuple of indices using the shape provided below.
+  optional TensorProto indices = 2;
```

Comment:
Note that we could generalize this and allow the user to choose to encode the indices in linearized form or directly: e.g., consider a 100x100 sparse matrix. A direct representation would be [ [row-indices], [column-indices] ] e.g., [ [0,1,2], [3,4,5] ] while a linearized form would be more compact as [ 0x100+3, 1x100+4, 2x100+5 ] = [3, 104, 205].

Reply:
Yes, then we need to have checker being aware of this.

Reply:
In linearized case, the rank is 1 and shape is [NNZ]. In the other case, the rank is 2, and the shape is either [NNZ, dims] or [dims, NNZ] (we need to pick one of these). Note that the values are stored separately (as they may have a different type) from the indices.

Reply:
I'd prefer `[NNZ, dims]` for more compact representation in extremely sparse cases.

Reply:
I'd prefer `[NNZ, dims]`. In general, sparse matrices are used to store `data points` (a collection of sparse feature vectors). I'd imagine that most data transformations would want to access the coordinates of one `non-zero` at the same time.

Reply:
Yes, conceptually [NNZ, dims] is more intuitive. I think TF does this. However, the MKL sparse library, for example, requires row-indices as a separate array and column-indices as a separate array (see: https://software.intel.com/en-us/mkl-developer-reference-c-sparse-blas-coordinate-matrix-storage-format ), which can be generated easily if we use [dims, NNZ]. So, it doesn't look like there is one uniform standard anyway.

Reply:
PyTorch sparse constructors (at Python level) seem to take a list of lists in [dims,NNZ] ordering: see: https://pytorch.org/docs/stable/sparse.html

Reply:
Let's only consider 2-D tensors for simplicity. Right, there is no uniform standard, but having `[NNZ, dims]` is enough to cover all cases. As COO format is not efficient for matrix computation, runtime will have to do a conversion (to, e.g., CSR) by sorting those non-zeros according to either row indexes and column indexes. Because the sorting will be something like `sorting those 1-D index vectors by their first/second value`, I feel using `[NNZ, dims]` may lead to better cache locality.

On the other hand, when producing sparse tensors to COO format using `OneHotEncoder`, we generally produce `[dims]` index vector at one time, so keep appending `[dims]` to `[NNZ, dims]`-output could also lead to a better cache locality.

Reply:
Another behavior I'd like to have is that `indices` should be sorted first by its first element, then by its second element, .., and finally its last element. This is very helpful when searching for a specified value in a sparse tensor.

Reply:
Yes, we should require that the indices should appear in sorted order. We need to clarify that in the documentation.